<html>
<head>
    <title>Thousand Genomes Variant Set Demo</title>
    <script src="https://code.jquery.com/jquery-1.12.3.min.js"   integrity="sha256-aaODHAgvwQW1bFOGXMeX+pC4PZIPsvn2h1sArYOhgXQ="   crossorigin="anonymous"></script>
</head>
<body style="padding-left: 0%; padding-right: 0%; font-family: sans-serif; margin: 0; padding: 0">

<h2>Thousand Genomes Variant Set Demo</h2>
<p>This is a glimpse at the data presented by the Thousand Genomes variants 
accessed via the GA4GH Genomics API. There are variant calls for <span id="length"></span> samples that are being accessed via HTTP.

<p>In this variant set, each variant record contains the calls made against it, i.e. 
  the individual genetic variations. In the visualization below, each variant is given a row  
  and the calls made against it are the columns, similar to the format of VCF. This gives us an overview of 
  the data that does <em>not</em> map directly to genomic position, although ordered by it.

<p>Each call for the variant is drawn as a single pixel. The pixels that are <span style="background: red; color: white; padding: 3px">red</span>
  represent a call on the first haplotype, <span style="background: green; color: white; padding: 3px">green</span> on the second, and <span style="background: yellow; padding: 3px">yellow</span> for both. We are looking at the beginning region of the first chromosome.
  </div>
  <p>
    <div style="width: 100%;" id="container">
      <p style="margin-left: 180px"><em>Calls</em></p>
      <span style="position: absolute; left: 5%; top: 550px"><em>Variants</em><br/><small></span>
        <div style="width: 100%; height: 400px; overflow: none">
  <canvas id="c" height="400" width="300" style="background: #333; margin-left: 15%">Your browser does not support canvas</canvas>
  </div>
</div>
  <div class="fold">
    <p>There are interesting features in these data that we can observe visually. <img src="1kg01.png" align="left" style="width: 25%; padding: 2%">It appears that for some variants, nearly all 
      samples received a positive call on both haplotypes. We can inspect these shared variants by 
      selecting the first variant observed in more than <span class="threshold">NaN</span> of the population. Here is one of those variants in GA4GH format:
    
    <p><pre id="shared" style="overflow: scroll">Well.. will be...</pre>
    
    <p>The long string identifiers are internal to the server instance and can be <a href="#" id="getvar" target="_new">accessed directly</a>. Some variants are named in the dbSNP database <a href="#"id="dbsnp">here</a>.
    
     <h2>Variant annotations</h2>
     
    <p>The Thousand Genomes project, when generating this dataset, calculated allele frequencies for each variant, which are available in the 'info' field of the record. We would expect them to be high for the variants that appear as yellow bands. For this variant the allele frequencies were: <pre><span id="af">still searching...</span></pre>
    
    <p>This variant was labeled <em><span id="type"></span></em> by the data preparers. GA4GH provides a variant annotation protocol that can be used to access Thousand Genomes functional annotations, which go beyond what is included in the basic release. By performing a search for the same region in the variant annotation set we find this variant has been tagged as '<span id="anntype"></span>'.
    
    <p>So far, this script has analyzed <span id="nvars">NaN:)</span> variants and <span id="sharedlen"></span> have been found to be shared by more than <span class="threshold"></span> participants. These variants span the 
      range <span class="min">0</span> to <span class="max">0</span> in genomic coordinates.


      <h2>Genetic features</h2>
      
    <p>Using the Sequence Annotations protocol of the genomics API we can learn more about the site of a variant. In this implementation the gencode genes annotations provide features for regions of DNA. So far the following genes have been observed:
      
      <p><span id="features"></span></p>


      <p>To learn more about the GA4GH and the Genomics API please visit <a href="http://github.com/ga4gh/" target="_new">github</a>.
        
<script type="text/javascript">
var baseurl = "http://1kgenomes.ga4gh.org/";
var c = document.getElementById("c");
var l = 300;//window.innerWidth;
var threshold = 0.8 * l;
c.width = l;
var url = "variants/search"
function getVariants(query, callback) {
  return $.ajax(baseurl + "variants/search", {
      data : JSON.stringify(query),
      contentType : 'application/json',
      type : 'POST',
      success: callback,
      dataType: 'json'
  });
}

function getReads(query, callback) {
  return $.ajax(baseurl + "reads/search", {
      data : JSON.stringify(query),
      contentType : 'application/json',
      type : 'POST',
      success: callback,
      dataType: 'json'
  });
}

function getFeatures(query, callback) {
  return $.ajax(baseurl + "features/search", {
      data : JSON.stringify(query),
      contentType : 'application/json',
      type : 'POST',
      success: callback,
      dataType: 'json'
  });
}

function getVariantAnnotations(query, callback) {
  return $.ajax(baseurl + "variantannotations/search", {
      data : JSON.stringify(query),
      contentType : 'application/json',
      type : 'POST',
      success: callback,
      dataType: 'json'
  });
}

var n = 0;

function reset() {
  n = 0;
  pageVariants(q);
  
}

function percent(n) {
  return Math.floor(1000 * n)  / 10;
}

var genes = {};

function updateVariants(res) {
  n += res.variants.length;

  res.variants.map(function(v, i) {
    window.setTimeout(function() {
      addVariant(v);
    }, 100 * i)
  });
  $("#length").html(Object.keys(calls).length);
  $("#nvars").html(Object.keys(variants).length);
  $(".threshold").html(threshold);
  $(".min").html(min);
  $(".max").html(max);
  getFeatures({"featureSetId": "WyIxa2dlbm9tZXMiLCJnZW5jb2RlX3YyNGxpZnQzNyJd", "start": 0, "end": max + 1, "parentId": "", "featureTypes": [], "referenceName": "chr1"}, function(res) {
    res.features.forEach(function(feature) {
      genes[feature.attributes.vals.gene_name] = true;
      var html = "<li>";
      html += Object.keys(genes).join("</li><li> ")
      html += "</li>"
      $("#features").html(html);
    })
  })
  $("#sharedlen").html(shared.length);
  if (shared.length > 0) {

    
    getVariantAnnotations({"variantAnnotationSetId": "WyIxa2dlbm9tZXMiLCJ2cyIsImZ1bmN0aW9uYWxfYW5ub3RhdGlvbiIsImZ1bmN0aW9uYWxfYW5ub3RhdGlvbiJd", "start": shared[0].start, "end": shared[0].end, "referenceName": "1"}, function(res) {
      $("#anntype").html(res.variantAnnotations[0].transcriptEffects[0].effects[0].term)
    })
    $("#getvar").attr("href", baseurl + "/variants/" + shared[0].id)
    var copy = JSON.parse(JSON.stringify(shared[0]));
    delete copy.calls;
    delete copy.info;
    
    $("#shared").html(JSON.stringify(copy, null, 2));
    $("#dbsnp").attr("href", "http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?searchType=adhoc_search&type=rs&rs=" + copy.names[0]);
    $("#dbsnp").html(copy.names[0]);
    $("#type").html(shared[0].info.VT[0]);
    var html = "";
    html += percent(shared[0].info.AFR_AF[0])
    html += "%, "
    html += percent(shared[0].info.EAS_AF[0])
    html += "%, "
    html += percent(shared[0].info.AMR_AF[0])
    html += "%, "
    html += percent(shared[0].info.EUR_AF[0])
    html += "%, "
    html += percent(shared[0].info.SAS_AF[0])
    html += "% "
    $("#af").html(html);
  }
}

function pageVariants(query) {
  // set start to equal end and split multiple requests
  var d = (new Date()).getTime();
  getVariants(query, function(res) {
    updateVariants(res);
    console.log("Got " + res.variants.length + " variants in " + ((new Date()).getTime() - d))
    d = (new Date()).getTime();
    if (res.nextPageToken) {
      var nq = query;
      nq.pageToken = res.nextPageToken;
      nq.pageSize = 5;
      setTimeout(function(){ pageVariants(nq); }, 10);
    } else {
      // assuming we are served in order
      console.log("done");
    }
  })
}


var calls = {};
var initCalls = true; // assume same calls in each

var x = 0;
var min = Infinity;
var max = -1;

var shared = [];

var variants = {};
var ncalled = {};
function addVariant(variant) {
  
  // draw two pixels for each call
  x += 1;
  if (variant.start < min) {
    min = variant.start;
  }
  
  if (variant.start > max) {
    max = variant.start;
  }
  
  // do we need to update the callset map?
  var l = Object.keys(calls).length;


  if (typeof variants[variant.id] === "undefined") {
    variants[variant.id] = Object.keys(variants).length;
  }
  
  var callsLength = Object.keys(calls).length;
  var csIds = variant.calls.map(function(call, i) {
    if (typeof calls[call.callSetId] === "undefined") {
      calls[call.callSetId] = callsLength + i;
    }
  });
  
  var ctx = c.getContext("2d");
  ctx.webkitImageSmoothingEnabled = false;
  ctx.imageSmoothingEnabled = false;
  var a = 255;
  var ncalledboth = 0;
  variant.calls.forEach(function(call, i) {
    // draw r = first haplotype
    // draw b = second haplotype
    if (typeof ncalled[variant.id] === "undefined") {
      ncalled[variant.id] = 0;
    }
    ncalled[variant.id] += call.genotype[1] * call.genotype[0];
    
    var g = call.genotype[1] * 255;
    var r = call.genotype[0] * 255;
    var b = 0;
    ctx.fillStyle = "rgba("+r+","+g+","+b+","+(a/255)+")";
    ctx.fillRect( calls[call.callSetId], variants[variant.id], 1, 1 );
  });
  if (ncalled[variant.id] > threshold && shared.indexOf(variant) == -1) {
    shared.push(variant);
  }
}

var cwidth = document.getElementById("container").offsetWidth;


var callSetIds = [];

function splitCall(n, leng) {
  for (var i = 0; i < n; i ++) {
    console.log(leng);
    console.log(n);
    var frac = leng / n;
    console.log(frac);
    console.log('slice in')
    console.log(i * frac)
    console.log('slice out')
    console.log((i + 1) * frac);

    pageVariants({
        "variantSetId": "WyIxa2dlbm9tZXMiLCJ2cyIsInJlbGVhc2UiXQ",
        "start": 0,
        "end": 1231132,
        "referenceName": "1",
        "callSetIds": callSetIds.slice(i * frac, (i + 1) * frac),
      "pageSize": 1
      });
  }
}
var callSetIds = [];
getVariants({"variantSetId": "WyIxa2dlbm9tZXMiLCJ2cyIsInJlbGVhc2UiXQ",
      "start": 0,
      "end": 1231132,
      "referenceName": "1",
      "pageSize": 1
    }, function(res) {
  callSetIds = res.variants[0].calls.map(function(call) {
    return call.callSetId;
  });
  console.log(callSetIds);
  
  splitCall(10, l);
})




  
// populate an interface to make a selection of callsetids
// and reference name

</script>